### ISSUE TITLE
[BUG] Docker: `UnboundLocalError: cannot access local variable 'data' where it is not associated with a value`

### ISSUE THREAD

**Issue by goraje** [dashingsoft/pyarmor#1528]
## Problem description

I am using PyArmor 8.3.10 (group license) on a M1 MacBook Pro as the host machine and the authentication service starts properly via `pyarmor-auth`. However, once I try to register the Docker container with `pyarmor reg` after copying the device file to the container I get the following error:
```
root@2931e6620257:/# pyarmor reg pyarmor-device-regfile-6005.6.zip
INFO     Python 3.11.6
INFO     Pyarmor 8.3.10 (trial), 000000, non-profits
INFO     Platform linux.aarch64
INFO     register "pyarmor-device-regfile-6005.6.zip"
ERROR    could not get docker host machine id
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/register.py", line 204, in _get_docker_hostname
    machid = (flag + data).decode('utf-8')
                     ^^^^
UnboundLocalError: cannot access local variable 'data' where it is not associated with a value
ERROR    could not get docker host machine id
```

## Steps to reproduce

- register a new host machine
```
pyarmor reg -g 6
pyarmor reg -g 6 pyarmor-regfile-6005.zip
```
`pyarmor` command output (redacted):
```
INFO     Python 3.10.10
INFO     Pyarmor 8.3.10 (group), 006005, ******
INFO     Platform darwin.aarch64
usage: pyarmor [-h] [-v] [-q] [-d]  ...
```
- start the authentication service on the host
```
pyarmor-auth pyarmor-device-regfile-6005.6.zip
```
`pyarmor-auth` output:
```
2023-10-12 14:55:20,031: work path: /Users/*****/.pyarmor/docker
2023-10-12 14:55:20,032: register "pyarmor-device-regfile-6005.6.zip"
2023-10-12 14:55:20,041: listen container auth request on 0.0.0.0:29092
```
- start a new Docker container
```
docker run -it --rm --name pyarmor-docker --add-host=host.docker.internal:host-gateway python:3.11 bash
```
- copy the license file to the container
```
docker cp pyarmor-device-regfile-6005.6.zip pyarmor-docker:/
```
- install Pyarmor in the Docker container and run the registration
```
pip install pyarmor
...
Installing collected packages: pyarmor.cli.core, pyarmor
Successfully installed pyarmor-8.3.10 pyarmor.cli.core-4.3.6

pyarmor reg pyarmor-device-regfile-6005.6.zip
```
`pyarmor-reg` output:
```
INFO     Python 3.11.6
INFO     Pyarmor 8.3.10 (trial), 000000, non-profits
INFO     Platform linux.aarch64
INFO     register "pyarmor-device-regfile-6005.6.zip"
ERROR    could not get docker host machine id
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/pyarmor/cli/register.py", line 204, in _get_docker_hostname
    machid = (flag + data).decode('utf-8')
                     ^^^^
UnboundLocalError: cannot access local variable 'data' where it is not associated with a value
ERROR    could not get docker host machine id
```

## Temporary "fix"

I have examined the part of the code mentioned in the `pyarmor-reg` error and I have noticed that the conditional `if` statement is never fulfilled in my case. If I change the following, original code:
```python
while True:
    flag = s.recv(1)
    if flag in bytes(MACHFLAGS):
        data = s.recv(32)
    machid = (flag + data).decode('utf-8')
``` 
to
```python
while True:
    flag = s.recv(1)
    data = s.recv(32)
    print("FLAG", flag) # for debug of the issue only
    print("DATA", data) # for debug of the issue only
    # if flag in bytes(MACHFLAGS):
    #    data = s.recv(32)
    machid = (flag + data).decode('utf-8')
``` 
everything seems to work as intended.

Additionally the contents of `.pyarmor/group/pyarmor-group-device.6` on the host reads as follows:
```
cat .pyarmor/group/pyarmor-group-device.6
# Generated by Pyarmor 8.3.10, 2023-10-12T13:37:07.084484
host: ***********
system: Darwin (Darwin Kernel Version 22.6.0: Fri Sep 15 13:41:28 PDT 2023; root:xnu-8796.141.3.700.8~1/RELEASE_ARM64_T6000)
machine: l98a1e7f27fc612f428043fb2edb27122
```
which is the same as
```
python
>>> from pyarmor.cli.core import Pytransform3
>>> print(Pytransform3.get_hd_info(21))
b'l98a1e7f27fc612f428043fb2edb27122'
```
and the values printed out  for `flag` and `data` in the modified code are as follows:
```
FLAG b'l'
DATA b'98a1e7f27fc612f428043fb2edb27122'
```
which together forms the machine ID of the host. It seems that for some reason the conditional clause that is present in the original code prevents it from working properly on my setup.

**Comment by tagnelli** [comment]
Hello, 
I noticed the same issue while running pyarmor inside a **Windows** docker container (isolation process on a Windows docker host) and while your temporary fix seems to let pyarmor register "correctly", when I try to do any kind of `pyarmor gen foo.py`, it doesn't work and trigger another error : 
```
PS C:\pyarmor> pyarmor -d gen foo.py
INFO     Python 3.11.6
INFO     Pyarmor 8.3.10 (group), 006083, non-profits
INFO     Platform windows.x86_64
DEBUG    native platform windows.amd64
DEBUG    home path: C:\Users\ContainerAdministrator\.pyarmor
DEBUG    command options: {'no_runtime': False, 'inputs': ['foo.py']}
DEBUG    install plugin: CodesignPlugin
DEBUG    install plugin: DarwinUniversalPlugin
INFO     search inputs ...
INFO     find script foo.py
INFO     find 1 top resources
gethostbyname: No error
DEBUG    unknown error, please check pyarmor.error.log
ERROR    invalid license token, try to run `pyarmor reg` to register license again
```

Are you in the same situation ? 

I have a production deployment date of this tool next Monday and I'm starting to think it's not gonna end well.

**Comment by goraje** [comment]
Hi, unfortunately it seems that you're right. I am in fact encountering exactly the same problem. I was happy with the initial success and didn't think things could go wrong from there. My output for reference:
```
pyarmor gen foo.py
INFO     Python 3.11.6
INFO     Pyarmor 8.3.10 (group), 006005, ******
INFO     Platform linux.aarch64
INFO     search inputs ...
INFO     find script foo.py
INFO     find 1 top resources
ERROR    invalid license token, try to run `pyarmor reg` to register license again
```

**Comment by jondy** [comment]
Try this patch for quick workaround
0b68e9ec78139a1e9b00d1cc026d5f8c52bf1834

**Comment by jondy** [comment]
Fixed in 8.3.11

**Comment by goraje** [comment]
Hi, I'm still getting an error although the communication between the authentication service and the container now work correctly.
```
pyarmor reg pyarmor-device-regfile-6005.6.zip
INFO     Python 3.11.6
INFO     Pyarmor 8.3.11 (trial), 000000, non-profits
INFO     Platform linux.aarch64
INFO     register "pyarmor-device-regfile-6005.6.zip"
INFO     machine id in group license: l98a1e7f27fc612f428043fb2edb27122
INFO     got machine id: lcdf0525b9b9521660872d6e0cd383062
INFO     got machine id: icdf0525b9b9521660872d6e0cd383062
INFO     got machine id: k946887efda4672de86115a564e06f8c7
INFO     got machine id: g946887efda4672de86115a564e06f8c7
INFO     got machine id: b3bd3014eb8fc9f71d02ea25e679320a4
INFO     no machine id matchs this group license
INFO     take this machine as docker container, and connect to docker host for authentication...
INFO     could not get docker host machine id

if this machine is docker container, please run command `pyarmor-auth` in docker host, and try it again

otherwise please generate new group device license for this machine

more information please check section "using group license" in documentation "how-to register" guide

ERROR    this group device license is not for this machine
```
I think this may be related to the fact that on both macOS and Windows the host system is not the actual "host" for the Docker containers. On both these OSes Docker is being run on a separate Linux machine that has its connection piped to the host and as such the machine ID's that Pyarmor is retrieving are not those of the macOS/Windows host but rather those of the Linux VM.

**Comment by tagnelli** [comment]
Hey, same issue here on Windows.
```
PS C:\pyarmor> pyarmor reg pyarmor-device-regfile-6083.5.zip
INFO     Python 3.11.6
INFO     Pyarmor 8.3.11 (trial), 000000, non-profits
INFO     Platform windows.x86_64
INFO     register "pyarmor-device-regfile-6083.5.zip"
INFO     machine id in group license: l5e241d78c365be6ceeb866236dd9b9c3
INFO     got machine id: l3cae7163c584dc46857cfd7234290c13
INFO     got machine id: i3cae7163c584dc46857cfd7234290c13
INFO     got machine id: k4ec9a0aeee009d88dfda21110413eaca
INFO     got machine id: g4ec9a0aeee009d88dfda21110413eaca
INFO     got machine id: b5f427f73248ef026d703425f3bc0b7f7
INFO     no machine id matchs this group license
INFO     take this machine as docker container, and connect to docker host for authentication...
INFO     could not get docker host machine id

if this machine is docker container, please run command `pyarmor-auth` in docker host, and try it again

otherwise please generate new group device license for this machine

more information please check section "using group license" in documentation "how-to register" guide

ERROR    this group device license is not for this machine
```
However, as we are running our docker container in isolation process on a Windows docker host and with a Windows based image, Linux is totally out of the equation in my case.
BUT, `host.docker.internal` doesn't resolve from within the container. So I have to run the following command in the container before trying to run `pyarmor reg`, while having `pyarmor-auth device.regfile` running on the docker hos :

```
$(Get-Content -Path C:\Python\Lib\site-packages\pyarmor\cli\register.py -Raw).Replace("host = 'host.docker.internal'","host = 'ip.of.my.docker.host'") | Set-Content -Path C:\Python\Lib\site-packages\pyarmor\cli\register.py
```
Which basically replaces `host.docker.internal` with the IP of my docker host. _(I also had to open port 29092 on my docker host)_

**Comment by jondy** [comment]
While the first reports is error `UnboundLocalError`, 8.3.11 is just fixing this problem.

The new issue is that machine id is not for this machine, there are lot of issues about this and there are many ways to debug, please first check these issues.

Also check opened issue #1523, it may same as this case.

I'll close this issue and don't track it.

**Comment by goraje** [comment]
Hi, unfortunately the proposed fix is not working. This is related directly to the patch that was merged to 8.3.11 [0b68e9ec7](https://github.com/dashingsoft/pyarmor/commit/0b68e9ec78139a1e9b00d1cc026d5f8c52bf1834). The conversion `bytes([x - 87 for x in MACHFLAGS])` is invalid because it is called on negative integers. This can be confirmed by running with the `--debug` flag.
```
INFO     Python 3.11.6
INFO     Pyarmor 8.3.11 (trial), 000000, non-profits
INFO     Platform linux.aarch64
DEBUG    native platform linux.aarch64
DEBUG    home path: /root/.pyarmor
INFO     register "pyarmor-device-regfile-6005.6.zip"
DEBUG    extracting license.lic
DEBUG    extracting .pyarmor_capsule.zip
INFO     machine id in group license: l98a1e7f27fc612f428043fb2edb27122
INFO     got machine id: l5bd8dfd8ad4f100902d43e6e0e8025be
INFO     got machine id: i5bd8dfd8ad4f100902d43e6e0e8025be
INFO     got machine id: k58444ccf00d89f5aa28e95778267b4d4
INFO     got machine id: g58444ccf00d89f5aa28e95778267b4d4
INFO     got machine id: b812ceb5191d569f7f09d136d173e5801
INFO     no machine id matchs this group license
INFO     take this machine as docker container, and connect to docker host for authentication...
DEBUG    host.docker.internal:29092:bytes must be in range(0, 256)
INFO     could not get docker host machine id

if this machine is docker container, please run command `pyarmor-auth` in docker host, and try it again

otherwise please generate new group device license for this machine

more information please check section "using group license" in documentation "how-to register" guide

ERROR    this group device license is not for this machine
```

**Comment by jondy** [comment]
Sorry, I made a mistake.

It should be `if flag in bytes([(x - 10) for x in MACHFLAGS]):`

**Comment by jondy** [comment]
Fixed in 8.4.0
